Test failures for cluster code with current numpy master

PR #1186 has a new matrix entry testing dipy against current numpy and scipy master builds, to make sure that we don't distribute stuff that will break on the next release of numpy / scipy.

This finds some new test failures on the cluster code, which appear to be because of a change in the way that numpy compares arrays.  I explored a little, and realized that I didn't understand the intention behind the tests, and in particular, whether the tests were really intended to test the behavior of `np.array(clusters)`.  Test failures are of the form:

```
======================================================================
FAIL: dipy.segment.tests.test_clustering.test_cluster_getitem
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/nipy/dipy/venv/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/nipy/dipy/venv/lib/python3.5/site-packages/dipy/segment/tests/test_clustering.py", line 120, in test_cluster_getitem
    [data[indices[i]] for i in advanced_indices])
  File "/home/travis/build/nipy/dipy/venv/lib/python3.5/site-packages/numpy/testing/utils.py", line 852, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "/home/travis/build/nipy/dipy/venv/lib/python3.5/site-packages/numpy/testing/utils.py", line 776, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not equal
(mismatch 100.0%)
 x: array([ array([[  0.,   1.,   2.],
       [  3.,   4.,   5.],
       [  6.,   7.,   8.],...
 y: array([ array([[  0.,   1.,   2.],
       [  3.,   4.,   5.],
       [  6.,   7.,   8.],...
```

https://travis-ci.org/nipy/dipy/jobs/208736688#L2742

You should be able to replicate the test failures by first doing:

```
pip install -U --pre -f https://7933911d6844c6c53a7d-47bd50c35cd79bd838daf386af554a83.ssl.cf2.rackcdn.com numpy
```

--- a/dipy/segment/tests/test_clustering.py
+++ b/dipy/segment/tests/test_clustering.py
@@ -116,7 +116,7 @@ def test_cluster_getitem():
         assert_array_equal(cluster[i], data[indices[i]])
 
     # Test advanced indexing
-    assert_array_equal(cluster[advanced_indices],
+    assert_arrays_equal(cluster[advanced_indices],
                        [data[indices[i]] for i in advanced_indices])
 
     # Test index out of bounds
@@ -245,8 +245,8 @@ def test_cluster_centroid_getitem():
         assert_array_equal(cluster[i], data[indices[i]])
 
     # Test advanced indexing
-    assert_array_equal(cluster[advanced_indices],
-                       [data[indices[i]] for i in advanced_indices])
+    assert_arrays_equal(cluster[advanced_indices],
+                        [data[indices[i]] for i in advanced_indices])
 
     # Test index out of bounds
     assert_raises(IndexError, cluster.__getitem__, len(cluster))
@@ -377,8 +377,8 @@ def test_cluster_map_iter():
 
     # Set refdata
     cluster_map.refdata = data
-    assert_array_equal(cluster_map, [[data[i] for i in cluster.indices]
-                                     for cluster in clusters])
+    for c1, c2 in zip(cluster_map, clusters):
+        assert_arrays_equal(c1, [data[i] for i in c2.indices])
 
     # Remove refdata, i.e. back to indices
     cluster_map.refdata = None
@@ -403,8 +403,8 @@ def test_cluster_map_getitem():
         assert_equal(cluster_map[i], clusters[i])
 
     # Test advanced indexing
-    assert_array_equal(cluster_map[advanced_indices],
-                       [clusters[i] for i in advanced_indices])
+    assert_arrays_equal(cluster_map[advanced_indices],
+                        [clusters[i] for i in advanced_indices])
 
     # Test index out of bounds
     assert_raises(IndexError, cluster_map.__getitem__, len(clusters))
@@ -650,8 +650,8 @@ def test_cluster_map_centroid_iter():
 
     # Set refdata
     cluster_map.refdata = data
-    assert_array_equal(cluster_map, [[data[i] for i in cluster.indices]
-                                     for cluster in clusters])
+    for c1, c2 in zip(cluster_map, clusters):
+        assert_arrays_equal(c1, [data[i] for i in c2.indices])
 
 
 def test_cluster_map_centroid_getitem():
@@ -673,8 +673,8 @@ def test_cluster_map_centroid_getitem():
         assert_equal(cluster_map[i], clusters[i])
 
     # Test advanced indexing
-    assert_array_equal(cluster_map[advanced_indices],
-                       [clusters[i] for i in advanced_indices])
+    assert_arrays_equal(cluster_map[advanced_indices],
+                        [clusters[i] for i in advanced_indices])
 
     # Test index out of bounds
     assert_raises(IndexError, cluster_map.__getitem__, len(clusters))

